packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,226 @@
+    ///         can therefore not be relayed again.
+    mapping(bytes32 => bool) public successfulMessages;
+
+    /// @notice Nonce for the next message to be sent, without the message version applied. Use the
+    ///         messageNonce getter which will insert the message version into the nonce to give you
+    ///         the actual nonce to be used for the message.
+    uint240 internal msgNonce;
+
+    /// @notice Emitted whenever a message is sent to the other chain.
+    /// @param data Encoded data of the message that was sent.
+    event SentMessage(bytes data);
+
+    /// @notice Emitted whenever a message is successfully relayed on this chain.
+    /// @param msgHash Hash of the message that was relayed.
+    event RelayedMessage(bytes32 indexed msgHash);
+
+    /// @notice Emitted whenever a message fails to be relayed on this chain.
+    /// @param msgHash Hash of the message that failed to be relayed.
+    event FailedRelayedMessage(bytes32 indexed msgHash);
+
+    /// @notice Enforces that cross domain message sender and source are set. Reverts if not.
+    ///         This is leveraged to differentiate between 0 and nil at tstorage slots.
+    modifier notEntered() {
+        assembly {
+            if eq(tload(ENTERED_SLOT), 0) {
+                mstore(0x00, 0xbca35af6) // 0xbca35af6 is the 4-byte selector of "NotEntered()"
+                revert(0x1C, 0x04)
+            }
+        }
+        _;
+    }
+
+    /// @notice Retrieves the sender of the current cross domain message. If not entered, reverts.
+    /// @return _sender Address of the sender of the current cross domain message.
+    function crossDomainMessageSender() external view notEntered returns (address _sender) {
+        assembly {
+            _sender := tload(CROSS_DOMAIN_MESSAGE_SENDER_SLOT)
+        }
+    }
+
+    /// @notice Retrieves the source of the current cross domain message. If not entered, reverts.
+    /// @return _source Chain ID of the source of the current cross domain message.
+    function crossDomainMessageSource() external view notEntered returns (uint256 _source) {
+        assembly {
+            _source := tload(CROSS_DOMAIN_MESSAGE_SOURCE_SLOT)
+        }
+    }
+
+    /// @notice Sends a message to some target address on a destination chain. Note that if the call
+    ///         always reverts, then the message will be unrelayable, and any ETH sent will be
+    ///         permanently locked. The same will occur if the target on the other chain is
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message to trigger the target address with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        if (_destination == block.chainid) revert MessageDestinationSameChain(_destination);
+
```

Comment:
We want to be sure to add

```solidity
require(_target != Predeploys.L2_TO_L2_CROSSDOMAIN_MESSENGER);
```

but with custom errors instead of require